homeassistant/components/zwave_js/services.py

```diff
@@ -0,0 +1,160 @@
+"""Methods and classes related to executing Z-Wave commands and publishing these to hass."""
+import logging
+from typing import Optional, cast
+
+import voluptuous as vol
+from zwave_js_server.const import CommandClass
+from zwave_js_server.model.node import Node as ZwaveNode, get_value_id
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import DeviceRegistry
+from homeassistant.helpers.entity_registry import EntityRegistry
+
+from . import const
+from .entity import get_home_and_node_id_from_device_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ZWaveServices:
+    """Class that holds our services (Zwave Commands) that should be published to hass."""
+
+    def __init__(
+        self, hass: HomeAssistant, dev_reg: DeviceRegistry, ent_reg: EntityRegistry
+    ):
+        """Initialize with hass, device registry, and entity registry objects."""
+        self._hass = hass
+        self._dev_reg = dev_reg
+        self._ent_reg = ent_reg
+
+    @callback
+    def async_register(self) -> None:
+        """Register all our services."""
+        self._hass.services.async_register(
+            const.DOMAIN,
+            const.SERVICE_SET_CONFIG_VALUE,
+            self.async_set_config_value,
+            schema=vol.All(
+                {
+                    vol.Exclusive(ATTR_DEVICE_ID, "id"): cv.string,
+                    vol.Exclusive(ATTR_ENTITY_ID, "id"): cv.string,
+                    vol.Optional(const.ATTR_ENDPOINT): vol.Coerce(int),
+                    vol.Required(const.ATTR_CONFIG_PROPERTY): vol.Coerce(int),
+                    vol.Optional(const.ATTR_CONFIG_PROPERTY_KEY_NAME): cv.string,
+                    vol.Required(const.ATTR_CONFIG_VALUE): vol.Coerce(int),
```

Comment:
Some advanced usage to think about... it would be nice to allow both labels and integer values, for parameters that provide "options". Classic zwave forced you to use label. OZW improved this by allowing either a label or the integer value. For example, [this parameter](https://github.com/zwave-js/node-zwave-js/blob/53f9c91c4120fd1b4ceedaa90f67f80e89a5767d/packages/config/config/devices/0x0063/ge_14294_zw3005.json#L37-L61):

```json
		"3": {
			"label": "Night Light",
			"description": "Defines the behavior of the blue LED. Default is on when switch is off.",
			"valueSize": 1,
			"minValue": 0,
			"maxValue": 2,
			"defaultValue": 0,
			"readOnly": false,
			"writeOnly": false,
			"allowManualEntry": false,
			"options": [
				{
					"label": "LED on when switch is OFF",
					"value": 0
				},
				{
					"label": "LED on when switch is ON",
					"value": 1
				},
				{
					"label": "LED always off",
					"value": 2
				}
			]
		},
```

The HA dump shows:
```json
            {
              "endpoint": 0,
              "commandClass": 112,
              "commandClassName": "Configuration",
              "property": 3,
              "propertyName": "Night Light",
              "ccVersion": 1,
              "metadata": {
                "type": "number",
                "readable": true,
                "writeable": true,
                "valueSize": 1,
                "min": 0,
                "max": 2,
                "default": 0,
                "format": 0,
                "allowManualEntry": false,
                "states": {
                  "0": "LED on when switch is OFF",
                  "1": "LED on when switch is ON",
                  "2": "LED always off"
                },
                "label": "Night Light",
                "description": "Defines the behavior of the blue LED. Default is on when switch is off.",
                "isFromConfig": true
              },
              "value": 0
            },
```

With the zwave integration, I was forced to use `value: LED always off`. With OZW I could use either `value: 2` or `value: LED always off`.  So it would be nice to use either way here as well.

Reply:
@cgarwood thoughts on @kpine's
suggested language changes? I'd imagine we want to use the same language for the service that we use for the WS API as well as the library, so we should make a decision now and make sure we are conforming across the board

Reply:
One more thought about the language - while I understand the desire to make the transition easy for users, the reality is that ZWave JS works differently than OZW did. It's easy enough to explain what the different service parameters mean in the service definition, but I think we potentially add confusion by calling it one thing in the service but it being listed as something else in the dumps

Reply:
that's specifically about the `property_key_name` to `bitwise` and `property` to `parameter` changes that were proposed above

Reply:
I don't think a user-facing public API should reflect internal details of the zwave-js API. It's easy enough for HA to map them to the correct API. "parameter" is the standard terminology used in z-wave, not just OZW. "property" and "property_key_name" are obscure and only used by z-wave JS; they have no meaning outside of the context of the API.

A user shouldn't be required to look at dump file (and then learn how to format and read it) in order to use this service call. The HA frontend should use the similar terms so that a user could go from UI to service call. There is also the zwave-js device database. Ideally a user should be able to look at that and translate it to a service call. Example: https://devices.zwave-js.io/?jumpTo=0x000c:0x0203:0x0001:0.0

There's no mention of property/property key in the DB, but there are "parameters", "labels", "values", "bits" and "bitmask".

If I go to the Z-Wave Alliance DB, [everything](https://products.z-wavealliance.org/products/2957/configs) is "parameter".

Other ideas:

Require the parameter to include the partial identifier (using config file format):
```
parameter: "31[0x01]"
value: 1
```

Allow a parameter label instead of a number:
```
parameter: LED 1 Blink Status (bottom)
value: 1
```

Not sure if either of those are better than requiring a third parameter for the bitmask. I'm usually not a fan of using labels because they can change, some people like them. Looking at the DB it's easy to pick out the integer values.

Unrelated, thinking about future support:
- bulk parameters API, e.g. set all partials at the same time, (would require an list of parameters)
- setting the full parameter and skipping the partials (set parameter 31 to some full integer number)

Reply:
I like this approach with the parameter label. And that label should then be exactly what the config UI also shows so it's really easy to find out what to type for the user instead of digging into json files or other loops.

If a label changes, the user can easily see that in the config UI.